crates/auth/src/auth/security.rs

```diff
@@ -0,0 +1,194 @@
+use std::boxed::Box;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use axum::body::Body;
+use axum::http::header::HeaderValue;
+use axum::http::{Request, Response};
+use axum::response::Response as AxumResponse;
+use dashmap::DashMap;
+use tower::{Layer, Service};
+
+use crate::config::{RateLimitConfig, SecurityHeadersConfig};
+
+/// Rate limiting state shared between middleware instances
+#[derive(Clone)]
+struct RateLimitState {
+    config: RateLimitConfig,
+    counters: Arc<DashMap<String, (u32, std::time::Instant)>>,
+}
```

Comment:
not sure I understand the rationale for implementing ratelimiting here

this can be a service in front of the gateway no?

Reply:
But then that should either be the node or gateway's responsibility, all the auth server needs to do is authenticate